src/main/java/org/folio/consortia/service/impl/PublicationServiceImpl.java

```diff
@@ -0,0 +1,113 @@
+import org.folio.consortia.domain.dto.PublicationResponse;
+import org.folio.consortia.exception.PublicationException;
+import org.folio.consortia.service.HttpRequestService;
+import org.folio.consortia.service.PublicationService;
+import org.folio.consortia.service.TenantService;
+import org.folio.consortia.service.UserTenantService;
+import org.folio.spring.FolioExecutionContext;
+import org.folio.spring.FolioModuleMetadata;
+import org.folio.spring.scope.FolioExecutionContextSetter;
+import org.springframework.core.task.TaskExecutor;
+import org.springframework.stereotype.Service;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+
+@Service
+@Log4j2
+@AllArgsConstructor
+public class PublicationServiceImpl implements PublicationService {
+  private final TenantService tenantService;
+  private final UserTenantService userTenantService;
+  private final FolioExecutionContext folioExecutionContext;
+  private final FolioModuleMetadata folioModuleMetadata;
+  private final HttpRequestService httpRequestService;
+  private final TaskExecutor asyncTaskExecutor;
+
+  @Override
+  public PublicationResponse publishRequest(UUID consortiumId, PublicationRequest publication) {
+    FolioExecutionContext currentTenantContext = (FolioExecutionContext) folioExecutionContext.getInstance();
+    validatePublicationRequest(consortiumId, publication, currentTenantContext);
+
+    PublicationResponse savedPublication = persistPublicationRecord();
+
+    List<CompletableFuture<Object>> futures = new ArrayList<>();
+    for (String tenantId : publication.getTenants()) {
+      try (var context = new FolioExecutionContextSetter(prepareContextForTenant(tenantId, folioModuleMetadata, currentTenantContext))) {
+        CompletableFuture<Object> future = new CompletableFuture<>();
+        executeAsyncTask(publication, tenantId, future);
+        futures.add(future);
+      }
+
+    }
+    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+      .thenAccept(cf -> updatePublicationsStatus(futures, savedPublication));
+
+    return buildPublicationResponse(savedPublication.getId());
+  }
+
+  private void executeAsyncTask(PublicationRequest publication, String tenantId, CompletableFuture<Object> future) {
+    asyncTaskExecutor.execute(() -> {
+      try {
+        var response = httpRequestService.postRequest(publication.getUrl(), publication.getPayload());
+        if (response.getStatusCode().is2xxSuccessful()) {
+          log.info("executeAsyncTask:: successfully called {}", publication.getUrl());
+          future.complete(response.getBody());
+        }
+      } catch (Exception t) {
+        log.error("executeAsyncTask:: Error making request on tenant {}", tenantId, t);
+        future.completeExceptionally(t);
+      }
+    });
+  }
+
+  private PublicationResponse persistPublicationRecord() {
+    var pr = new PublicationResponse()
+      .id(UUID.randomUUID().toString())
```

Comment:
I think we should get id from PublicationRequest, because id in PublicationRequest and PublicationResponse should match, also make sure this this id included in all related logs